### PR TITLE
re-implement chaosctl based on ctrlserver

### PR DESCRIPTION
**Pull request by Hexilee**
Signed-off-by: hexilee <i@hexilee.me>

### What problem does this PR solve?

The PR #2192 brings us a ctrl server, now we are going to implement a new `chaosctl` based on it.

This PR is a part of #2243, without the `query` subcommand.

Problem Summary:

### What is changed and how it works?

Proposal: [rfc#40](https://github.com/chaos-mesh/rfcs/pull/40) 

What's Changed:

### Checklist

Tests

- [ ] Unit test
- [x] E2E test
- [x] Manual test (add detailed scripts or steps below)

Refer to [usage of chaosctl](https://github.com/chaos-mesh/chaos-mesh/blob/bef904904603b8bd81a8bc780bcbf58f3fa1d98c/pkg/chaosctl/README.md#detail-of-debug).

- [ ] No code

Side effects

- [ ] Breaking backward compatibility

### Release note 

```release-note
Please add a release note.
If you don't think this PR needs a release note then fill it with None.
```

### CODE REVIEW

**Comment by Hexilee** [comment]
Hi, I split a part of #2243 to reduce the workload of the review. This PR removes subcommand `query` compared with #2243. PTAL  @STRRL @YangKeao @cwen0

File: e2e-test/config.go [hunk 1]
```diff
@@ -38,14 +38,14 @@ func NewDefaultConfig() *Config {
 		ChartDir:             "/charts",
 		ManifestDir:          "/manifests",
 		Tag:                  "e2e",
-		ManagerImageRegistry: "localhost:5000",
```
Comment (YangKeao) [inline e2e-test/config.go]: If you'd like to modify these files, please submit another PR.

Comment (Hexilee) [reply e2e-test/config.go]: It seems you are reviewing an outdated commit, please refresh and continue.

File: pkg/ctrl/server/cgroup.go [hunk 2]
```diff
@@ -0,0 +1,119 @@
...
+	return cgroups, nil
+}
+
+// GetCgroup returns result of cat /proc/:pid/cgroup
+func (r *Resolver) GetCgroup(ctx context.Context, obj *v1.Pod, pid string) (string, error) {
+	cmd := fmt.Sprintf("cat /proc/%s/cgroup", pid)
+	return r.ExecBypass(ctx, obj, cmd, bpm.PidNS, bpm.MountNS)
+}
+
+// GetCPUQuota returns result of cat cat /sys/fs/cgroup/:cpuMountType/cpu.cfs_quota_us
```
Comment (WangXiangUSTC) [inline pkg/ctrl/server/cgroup.go:87]:
```suggestion
// GetCPUQuota returns result of cat /sys/fs/cgroup/:cpuMountType/cpu.cfs_quota_us
```

File: pkg/ctrl/server/fd.go [hunk 3]
```diff
@@ -13,22 +13,24 @@
...
 	"strings"
 
-	"github.com/chaos-mesh/chaos-mesh/pkg/ctrlserver/graph/model"
+	"github.com/chaos-mesh/chaos-mesh/pkg/bpm"
+	"github.com/chaos-mesh/chaos-mesh/pkg/ctrl/server/model"
 )
 
 // GetFdsOfProcess returns fd-target pairs
-func (r *Resolver) GetFdsOfProcess(ctx context.Context, process *model.Process) ([]*model.Fd, error) {
+func (r *Resolver) GetFdsOfProcess(ctx context.Context, process *model.Process) []*model.Fd {
```
Comment (WangXiangUSTC) [inline pkg/ctrl/server/fd.go:28]: why not return error?

File: test/integration_test/chaosctl/run.sh [hunk 4]
```diff
@@ -44,17 +47,120 @@ spec:
...
+  target: Request
+  port: 8081
+  path: "*"
+  delay: "10ms"
+  duration: "30s" # duration for the injected chaos experiment
+EOF
+kubectl apply -f delay.yaml
+
+echo "Checking chaosctl debug httpchaos"
+./chaosctl debug httpchaos web-show-http-delay
```
Comment (WangXiangUSTC) [inline test/integration_test/chaosctl/run.sh:90]: can we check the output of chaosctl?

File: pkg/chaosctl/cmd/physicalmachine.go [hunk 5]
```diff
@@ -16,17 +16,12 @@
...
-
-	"github.com/go-logr/logr"
 	"github.com/spf13/cobra"
 
 	"github.com/chaos-mesh/chaos-mesh/pkg/chaosctl/physicalmachine"
 )
 
-func NewPhysicalMachineCommand(logger logr.Logger) (*cobra.Command, error) {
-
+func NewPhysicalMachineCommand() (*cobra.Command, error) {
```
[the remark this answers is not in the document]
Comment (Hexilee) [reply pkg/chaosctl/cmd/physicalmachine.go:24]: Because the logger is only used to print errors, which is unfriendly for a cli application. I prefer return errors and print out by the pretty printer.

[the remark this answers is not in the document]
Comment (Hexilee) [reply pkg/chaosctl/cmd/physicalmachine.go:24]: Refer to https://github.com/chaos-mesh/chaos-mesh/pull/2243#issuecomment-952522490

File: pkg/chaosctl/debug/debug.go [hunk 6]
```diff
@@ -0,0 +1,30 @@
...
+	"context"
+
+	"github.com/chaos-mesh/chaos-mesh/pkg/chaosctl/common"
+	ctrlclient "github.com/chaos-mesh/chaos-mesh/pkg/ctrl/client"
+)
+
+type Debugger interface {
+	Run(ctx context.Context, namespace, chaosName string) ([]*common.ChaosResult, error)
+	List(ctx context.Context, namespace string) ([]string, error)
+}
```
[the remark this answers is not in the document]
Comment (Hexilee) [reply pkg/chaosctl/debug/debug.go:28]: Fine

File: pkg/chaosctl/debug/networkchaos.go [hunk 7]
```diff
@@ -0,0 +1,134 @@
...
+			result.Pods = append(result.Pods, podResult)
+		}
+
+		results = append(results, result)
+	}
+	return results, nil
+}
+
+// List list chaos
+func (d *networkDebugger) List(ctx context.Context, namespace string) ([]string, error) {
```
[the remark this answers is not in the document]
Comment (Hexilee) [reply pkg/chaosctl/debug/networkchaos.go:107]: The FieldTag of a struct can only be modified by reflection or code generator. I think the codebase is not large enough to use reflection or code generator.

File: pkg/ctrl/client/ctrlclient.go [hunk 8]
```diff
@@ -0,0 +1,56 @@
...
+
+	"github.com/hasura/go-graphql-client"
+)
+
+type CtrlClient struct {
+	QueryClient        *graphql.Client
+	SubscriptionClient *graphql.SubscriptionClient
+}
+
+func NewCtrlClient(ctx context.Context, url string) (*CtrlClient, error) {
```
[the remark this answers is not in the document]
Comment (Hexilee) [reply pkg/ctrl/client/ctrlclient.go:29]: Yeah

File: pkg/ctrl/server/cgroup.go (hunk 2 above)
Comment (Hexilee) [reply pkg/ctrl/server/cgroup.go:87]: typo.

File: pkg/ctrl/server/fd.go (hunk 3 above)
Comment (Hexilee) [reply pkg/ctrl/server/fd.go:28]: Because the errors often occur on some short-life process.

Comment (Hexilee) [reply pkg/ctrl/server/fd.go:28]: I will leave some comments to explain it.

File: test/integration_test/chaosctl/run.sh (hunk 4 above)
Comment (Hexilee) [reply test/integration_test/chaosctl/run.sh:90]: The output format is human-friendly and hard to read by machine.

File: pkg/ctrl/server/fd.go (hunk 3 above)
Comment (Hexilee) [reply pkg/ctrl/server/fd.go:28]: Left.

File: pkg/ctrl/server/cgroup.go (hunk 2 above)
Comment (Hexilee) [reply pkg/ctrl/server/cgroup.go]: Fixed

File: pkg/ctrl/client/ctrlclient.go (hunk 8 above)
[the remark this answers is not in the document]
Comment (Hexilee) [reply pkg/ctrl/client/ctrlclient.go]: Fixed

File: pkg/chaosctl/debug/debug.go (hunk 6 above)
[the remark this answers is not in the document]
Comment (Hexilee) [reply pkg/chaosctl/debug/debug.go]: Added

File: pkg/ctrl/server/cgroup.go [hunk 9]
```diff
@@ -0,0 +1,119 @@
...
+	v1 "k8s.io/api/core/v1"
+
+	"github.com/chaos-mesh/chaos-mesh/pkg/bpm"
+	"github.com/chaos-mesh/chaos-mesh/pkg/ctrl/server/model"
+)
+
+// GetCgroups returns result of cat /proc/cgroups
+func (r *Resolver) GetCgroups(ctx context.Context, obj *model.PodStressChaos) (*model.Cgroups, error) {
+	cmd := "cat /proc/cgroups"
+	raw, err := r.ExecBypass(ctx, obj.Pod, cmd, bpm.PidNS, bpm.MountNS)
```
Comment (WangXiangUSTC) [inline pkg/ctrl/server/cgroup.go:35]: how about adding a comment for what's the raw looks like?

File: pkg/ctrl/server/cgroup.go [hunk 10]
```diff
@@ -0,0 +1,119 @@
...
+	if obj.StressChaos.Spec.Stressors.CPUStressor == nil {
+		isCPU = false
+	}
+
+	if isCPU {
+		var cpuMountType string
+		if regexp.MustCompile("(cpu,cpuacct)").MatchString(string(raw)) {
+			cpuMountType = "cpu,cpuacct"
+		} else {
+			cpuMountType = "cpu"
```
Comment (WangXiangUSTC) [inline pkg/ctrl/server/cgroup.go:59]: what is the difference between these two types?

File: pkg/ctrl/server/cgroup.go (hunk 9 above)
Comment (WangXiangUSTC) [reply pkg/ctrl/server/cgroup.go:35]: Also for other cmds below, I think it's better to add some examples of the cmd's output, to make it easier to review

File: test/integration_test/chaosctl/run.sh (hunk 4 above)
Comment (WangXiangUSTC) [reply test/integration_test/chaosctl/run.sh:90]: refer to https://github.com/chaos-mesh/chaos-mesh/blob/21564576a61fb21eaf0ba7e60824231add9310bb/test/integration_test/jvm/run.sh#L78
just make sure the output contains some key information

File: pkg/ctrl/server/schema.resolvers.go [hunk 11]
```diff
@@ -980,8 +1014,34 @@ func (r *podStatusResolver) QosClass(ctx context.Context, obj *v1.PodStatus) (st
...
+	processes, err := r.Pod().Processes(ctx, obj.Pod)
+	if err != nil {
+		return nil, err
+	}
+
+	var processStress []*model.ProcessStress
+	for _, process := range processes {
+		cgroup, err := r.GetCgroup(ctx, obj.Pod, process.Pid)
+		if err != nil {
+			// ignore this process
```
Comment (WangXiangUSTC) [inline pkg/ctrl/server/schema.resolvers.go:1031]: how about adding a debug log

File: pkg/chaosctl/cmd/root.go [hunk 12]
```diff
@@ -42,44 +43,43 @@ Interacting with chaos mesh
...
-			"errorVerbose", fmt.Sprintf("%+v", err),
-		)
+		cm.PrettyPrint("failed to initialize cmd: ", 0, cm.Red)
+		cm.PrettyPrint("debug command: "+err.Error(), 1, cm.Red)
 		os.Exit(1)
 	}
 
-	physicalMachineCommand, err := NewPhysicalMachineCommand(rootLogger.WithName("cmd-physical-machine"))
+	rootCmd.Flags().StringVarP(&managerNamespace, "manager-namespace", "n", "chaos-testing", "the namespace chaos-controller-manager in")
+	rootCmd.Flags().StringVarP(&managerSvc, "manager-svc", "s", "chaos-mesh-controller-manager", "the service to chaos-controller-manager")
```
Comment (STRRL) [inline pkg/chaosctl/cmd/root.go:78]: It should use PersistentFlags here, and `--manager-namespace/-n` would be a conflict with `--namespace/-n` in command `debug` and `--node/-n` in command logs.

I usually install Chaos Mesh in kubernetes namespace `chaos-mesh` not `chaos-testing`, so when I want to declare the namespace, it not works.

**Review by STRRL — COMMENTED** [review]
Hi @Hexilee , PTAL

File: pkg/ctrl/server/cgroup.go (hunk 9 above)
Comment (Hexilee) [reply pkg/ctrl/server/cgroup.go:35]: OK

File: pkg/ctrl/server/cgroup.go (hunk 10 above)
Comment (Hexilee) [reply pkg/ctrl/server/cgroup.go:59]: The latter means cgroup does not support cpuacct sub-system, I left comments

File: pkg/ctrl/server/cgroup.go (hunk 9 above)
Comment (Hexilee) [reply pkg/ctrl/server/cgroup.go:54]: Comments left.

File: pkg/ctrl/server/cgroup.go (hunk 10 above)
Comment (Hexilee) [reply pkg/ctrl/server/cgroup.go]: Comments is added.

File: pkg/ctrl/server/schema.resolvers.go (hunk 11 above)
Comment (Hexilee) [reply pkg/ctrl/server/schema.resolvers.go:1032]: Printed this error out.

File: test/integration_test/chaosctl/run.sh (hunk 4 above)
Comment (Hexilee) [reply test/integration_test/chaosctl/run.sh]: Added some assertion.

File: pkg/chaosctl/cmd/debug.go [hunk 13]
```diff
@@ -68,104 +57,61 @@ Examples:
...
+}
+
+func debugResourceCommand(option *DebugOptions, chaosType string, debug debug.Debug) *cobra.Command {
+	return &cobra.Command{
+		Use:   fmt.Sprintf(`%s (CHAOSNAME) [-n NAMESPACE]`, chaosType),
+		Short: fmt.Sprintf(`Print the debug information for certain %s`, chaosType),
+		Long:  fmt.Sprintf(`Print the debug information for certain %s`, chaosType),
 		RunE: func(cmd *cobra.Command, args []string) error {
-			clientset, err := common.InitClientSet()
+			ctrlclient.DisableRuntimeErrorHandler()
```
[the remark this answers is not in the document]
Comment (Hexilee) [reply pkg/chaosctl/cmd/debug.go]: Now, we only call `ctrlclient.DisableRuntimeErrorHandler()` in the port forwarder as this function is used to prevent complaining from port forwarder.

File: pkg/chaosctl/debug/debug.go [hunk 14]
```diff
@@ -0,0 +1,33 @@
...
+import (
+	"context"
+
+	"github.com/chaos-mesh/chaos-mesh/pkg/chaosctl/common"
+	ctrlclient "github.com/chaos-mesh/chaos-mesh/pkg/ctrl/client"
+)
+
+type Debugger interface {
+	// Run to get chaos debug information
+	Run(ctx context.Context, namespace, chaosName string) ([]*common.ChaosResult, error)
```
[the remark this answers is not in the document]
Comment (Hexilee) [reply pkg/chaosctl/debug/debug.go:27]: Renamed to `Collect`.

File: pkg/chaosctl/debug/debug.go [hunk 15]
```diff
@@ -0,0 +1,33 @@
...
+
+	// List chaos
+	List(ctx context.Context, namespace string) ([]string, error)
```
[the remark this answers is not in the document]
Comment (Hexilee) [reply pkg/chaosctl/debug/debug.go]: Modified to "List chaos names to collect"

File: pkg/chaosctl/cmd/logs.go [hunk 16]
```diff
@@ -20,28 +20,33 @@ import (
...
-	o := &logsOptions{
-		logger: logger,
-	}
+type Component string
+
+const (
+	Manager   Component = "MANAGER"
+	Daemon    Component = "DAEMON"
+	Dashboard Component = "DASHBOARD"
+	DnsServer Component = "DNSSERVER"
```
[the remark this answers is not in the document]
Comment (Hexilee) [reply pkg/chaosctl/cmd/logs.go:44]: Used.

File: pkg/ctrl/client/ctrlclient.go [hunk 17]
```diff
@@ -0,0 +1,54 @@
...
+
+	"github.com/hasura/go-graphql-client"
+)
+
+type CtrlClient struct {
+	QueryClient        *graphql.Client
+	SubscriptionClient *graphql.SubscriptionClient
+}
+
+func NewCtrlClient(ctx context.Context, url string) *CtrlClient {
```
[the remark this answers is not in the document]
Comment (Hexilee) [reply pkg/ctrl/client/ctrlclient.go]: Removed

File: pkg/chaosctl/cmd/root.go (hunk 12 above)
Comment (Hexilee) [reply pkg/chaosctl/cmd/root.go]: Fixed, now the shorthand of `manager-namespace` is `-N`.

**Review by WangXiangUSTC — APPROVED** [review]
LGTM

**Review by STRRL — APPROVED** [review]
LGTM!
